modules/OFConnectionManager2/module/src/cxn_instance.h

```diff
@@ -0,0 +1,337 @@
+ * CXN_S_INIT                Connecting
+ * CXN_S_HANDSHAKING         TCP Connection established
+ * CXN_S_HANDSHAKE_COMPLETE  OpenFlow handshake (features) complete
+ * CXN_S_CLOSING             Connection broken, cleaning up
+ *                           outstanding operations
+ * CXN_S_CLOSED              Clean up complete;
+ *                           connection_t can be released
+ *
+ * Timeouts are 5 seconds for all states except HANDSHAKE_COMPLETE and CLOSED.
+ */
+
+/*
+ * each connection state is a tuple of
+ *   (value, id, human-readable name, timeout).
+ * the timeouts have units of milliseconds.
+ * a timeout of zero indicates that there is no timeout.
+ */
+#define CXN_S_DESCS                                                     \
+    cxn_s_desc(0, CXN_S_INIT, "Initialized", 5000)                      \
+    cxn_s_desc(1, CXN_S_HANDSHAKING, "Handshaking", 5000)               \
+    cxn_s_desc(2, CXN_S_HANDSHAKE_COMPLETE, "Handshake Complete", 0)    \
+    cxn_s_desc(3, CXN_S_CLOSING, "Closing", 5000)                       \
+    cxn_s_desc(4, CXN_S_CLOSED, "Closed", 0)                            \
+    cxn_s_desc(5, CXN_S_COUNT, "COUNT", 0) /* last value */
+
+typedef enum cxn_state_e {
+#define cxn_s_desc(_val, _id, _name, _timeout) _id = _val,
+    CXN_S_DESCS
+#undef cxn_s_desc
+} cxn_state_t;
+
+
+#define MAX_BUNDLES 4
+#define BUNDLE_ID_INVALID (-1)
+
+typedef struct bundle_s {
+    uint32_t id;    /* Supplied by controller when bundle is opened */
+    uint32_t count; /* Number of messages in bundle */
+    uint32_t allocated; /* Length of msgs array */
+    uint32_t bytes; /* Sum of message sizes */
+    uint16_t flags; /* Copied from flags field in bundle-open request */
+    uint8_t **msgs; /* Array of pointers to raw message data */
+} bundle_t;
+
+
+#define MAX_AUX_ID_DESC_LEN (4)
+
+/* Connection control block */
+typedef struct connection_s {
+    cxn_state_t state;
+    indigo_cxn_status_t status;
+
+    int active; /* Has this connection instance been configured? */
+    indigo_cxn_id_t cxn_id; /* For back tracking */
+
+    int sd; /* The socket descriptor */
+
+    /*
+     * The read buffer is used to buffer a single message.  There
+     * are two steps to the operation: Read the header, then use that
+     * to determine the number of bytes remaining in the message.
+     * Once a message is read in, it is processed and the read buffer
+     * is cleared.
+     */
+    uint8_t read_buffer[READ_BUFFER_SIZE];
+    int read_bytes; /* Number of bytes currently in read buffer */
+    int bytes_needed; /* Num bytes needed for next process step */
+
+    /* Write queue */
+    bigring_t *write_queue; /* Ringbuffer of pointers to message data */
+    int write_queue_head_offset; /* Bytes already sent out from head of write_queue */
+    int bytes_enqueued;     /* Total bytes queued */
+    int pkts_enqueued;      /* Total pkts queued */
+
+    /* Additional debug info */
+    int has_debug_counters;
```

Comment:
changed to bool.
connections that have not advanced to HANDSHAKING (cxn_try_to_connect did not succeed) don't have debug counters.